components/ntp_background_images/browser/ntp_sponsored_image_source.cc

```diff
@@ -52,137 +37,66 @@ void NTPSponsoredImageSource::StartDataRequest(
     GotDataCallback callback) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
-  const std::string path = URLDataSource::URLToRequestPath(url);
-  if (!IsValidPath(path)) {
-    content::GetUIThreadTaskRunner({})->PostTask(
-        FROM_HERE, base::BindOnce(std::move(callback),
-                                  scoped_refptr<base::RefCountedMemory>()));
-
-    return;
+  if (!background_images_service_) {
+    return DenyAccess(std::move(callback));
   }
 
-  base::FilePath image_file_path = GetLocalFilePathFor(path);
-  CHECK(!image_file_path.empty());
-  GetImageFile(image_file_path, std::move(callback));
-}
-
-void NTPSponsoredImageSource::GetImageFile(
-    const base::FilePath& image_file_path,
-    GotDataCallback callback) {
-  base::ThreadPool::PostTaskAndReplyWithResult(
-      FROM_HERE, {base::MayBlock()},
-      base::BindOnce(&ReadFileToString, image_file_path),
-      base::BindOnce(&NTPSponsoredImageSource::OnGotImageFile,
-                     weak_factory_.GetWeakPtr(), std::move(callback)));
-}
+  const NTPSponsoredImagesData* const images_data =
+      background_images_service_->GetSponsoredImagesData(
+          /*super_referral=*/false,
+          /*supports_rich_media=*/false);
+  if (!images_data) {
+    return DenyAccess(std::move(callback));
+  }
 
-void NTPSponsoredImageSource::OnGotImageFile(GotDataCallback callback,
-                                             std::optional<std::string> input) {
-  if (!input) {
-    return;
+  const base::FilePath request_path =
+      base::FilePath::FromUTF8Unsafe(URLToRequestPath(url));
+  const std::optional<base::FilePath> file_path =
+      MaybeGetFilePathForRequestPath(request_path, images_data->campaigns);
+  if (!file_path) {
+    return DenyAccess(std::move(callback));
   }
 
-  std::move(callback).Run(
-      new base::RefCountedBytes(base::as_byte_span(*input)));
+  AllowAccess(*file_path, std::move(callback));
 }
 
 std::string NTPSponsoredImageSource::GetMimeType(const GURL& url) {
-  const std::string path = URLDataSource::URLToRequestPath(url);
-  const base::FilePath file_path = base::FilePath::FromUTF8Unsafe(path);
-  if (file_path.MatchesExtension(FILE_PATH_LITERAL(".jpg")) ||
-      file_path.MatchesExtension(FILE_PATH_LITERAL(".jpeg"))) {
-    return "image/jpeg";
-  }
-  if (file_path.MatchesExtension(FILE_PATH_LITERAL(".png"))) {
-    return "image/png";
-  }
-  if (file_path.MatchesExtension(FILE_PATH_LITERAL(".webp"))) {
-    return "image/webp";
-  }
-  if (file_path.MatchesExtension(FILE_PATH_LITERAL(".avif"))) {
-    return "image/avif";
+  std::string mime_type;
+  const base::FilePath file_path =
+      base::FilePath::FromUTF8Unsafe(url.path_piece());
+  if (!file_path.empty()) {
+    net::GetWellKnownMimeTypeFromFile(file_path, &mime_type);
```

Comment:
this seems to be allowing more MIME types than previously? presumably we don't want to support arbitrary file extensions like svg.

Reply:
Like RichNTT, I made this change because ads-serve has these checks, your thoughts?

Reply:
i thought richNTT also only supported whitelisted MIME types?

Reply:
RichNTT has the same code as above.

Reply:
ok i guess this is fine if @tackley can confirm MIME validation is present server side for these assets

Reply:
still waiting on ^

Reply:
sorry for the delay, this one slipped through my list

@diracdeltas for NTT image uploads we explicitly validate on the extension - https://github.com/brave/ads-serve/blob/d18f09093b9464794dca71588dcde625c87db0f8/src/internal/internal.controller.ts#L103-L110.  

For consistency I've raised https://github.com/brave/ads-serve/issues/4775 to update this old code to use mime type lookup instead of a list of extensions. I'll also remove gif from the list as we've never used this.